distributed/protocol/core.py

```diff
@@ -0,0 +1,221 @@
+"""
+The distributed message protocol consists of the following parts:
+
+1.  The length of the header, stored as a uint32
+2.  The header, stored as msgpack.
+    If there are no fields in the header then we skip it entirely.
+3.  The payload, stored as possibly compressed msgpack
+4.  A sentinel value
+
+**Header**
+
+The Header contains the following fields:
+
+* **compression**: string, optional
+    One of the following: ``'snappy', 'lz4', 'zlib'`` or missing for None
+
+**Payload**
+
+The payload is any msgpack serializable value.  It may be compressed based
+on the header.
+
+**Sentinel**
+
+We often terminate each message with a sentinel value.  This happens
+outside of this module though and is not baked in.
+"""
+from __future__ import print_function, division, absolute_import
+
+from copy import deepcopy
+from functools import partial
+import logging
+
+try:
+    import pandas.msgpack as msgpack
+except ImportError:
+    import msgpack
+
+from toolz import identity, get_in, valmap
+
+from .compression import compressions, maybe_compress
+from .serialize import (serialize, deserialize, Serialize, Serialized,
+        to_serialize)
+from .utils import frame_split_size
+
+from ..utils import ignoring
+
+_deserialize = deserialize
+
+
+logger = logging.getLogger(__file__)
+
+
+def dumps(msg):
+    """ Transform Python message to bytestream suitable for communication """
+    try:
+        data = {}
+        # Only lists and dicts can contain serialized values
+        if isinstance(msg, (list, dict)):
+            msg, data = extract_serialize(msg)
+        small_header, small_payload = dumps_msgpack(msg)
+
+        if not data:  # fast path without serialized data
+            return small_header, small_payload
+
+        pre = {key: (value.header, value.frames)
+               for key, value in data.items()
+               if type(value) is Serialized}
+
+        data = {key: serialize(value.data)
+                     for key, value in data.items()
+                     if type(value) is Serialize}
+
+        header = {'headers': {},
+                  'keys': []}
+        out_frames = []
+
+        for key, (head, frames) in data.items():
+            if 'compression' not in head:
+                frames = frame_split_size(frames)
+                compression, frames = zip(*map(maybe_compress, frames))
+                head['compression'] = compression
+            head['lengths'] = list(map(len, frames))
+            header['headers'][key] = head
+            header['keys'].append(key)
+            out_frames.extend(frames)
+
+        for key, (head, frames) in pre.items():
+            head['lengths'] = list(map(len, frames))
+            header['headers'][key] = head
+            header['keys'].append(key)
+            out_frames.extend(frames)
+
+        out_frames = [bytes(f) for f in out_frames]
```

Comment:
This possible copy should become unnecessary if we can get something like https://github.com/tornadoweb/tornado/pull/1691 in to tornado